pilot/tools/adscagent/adscagent.go

```diff
@@ -0,0 +1,182 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This tool simulate envoy sidecar gRPC call to get config,
+//
+// Usage:
+//
+// First, you can either manually expose pilot gRPC port, run locally or connect to a real XDS server.
+//
+// * By port-forward existing pilot - same can be done with Skaffold:
+//
+// ```bash
+// kubectl port-forward $(kubectl get pod -l istio=pilot -o jsonpath={.items[0].metadata.name} -n istio-system) -n istio-system 15010
+// ```
+//
+// * Or run local pilot using the default .kube/config or $KUBECONFIG:
+//
+//```bash
+// pilot-discovery discovery
+// ```
+//
+// To get LDS or CDS, use -type lds or -type cds, and provide the pod labels. For example:
+//
+// ```bash
+// go run pilot_cli.go
+// ```
+// ```
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	gogojsonpb "github.com/gogo/protobuf/jsonpb"
+	golangjsonpb "github.com/golang/protobuf/jsonpb"
+	"istio.io/api/mesh/v1alpha1"
+	//"io/ioutil"
+	//"istio.io/istio/pilot/pkg/model"
+	//"istio.io/istio/pkg/adsc"
+	istio_agent "istio.io/istio/pkg/istio-agent"
+	"istio.io/istio/pkg/security"
+
+	//	"istio.io/istio/pkg/util/gogoprotomarshal"
+
+	_ "net/http"
+	_ "net/http/pprof"
+
+	"istio.io/pkg/env"
+	"istio.io/pkg/log"
+)
+
+var (
+	// Same as normal agent
+	namespace = env.RegisterStringVar("POD_NAMESPACE", "dev", "Pod namespace")
+
+	// TODO: labels, PROXY_CONFIG
+
+	proxyName  = flag.String("name", "default", "Pod name")
+	pilotURL   = flag.String("xds", "localhost:15010", "xds server address.")
```

Comment:
We are using `xds-address` for this in _istioctl_.  We should use the same flag name.

Reply:
I was thinking to use "discoveryAddress" or even a PROXY_CONFIG env variable..
Maybe istioctl should do the same...

Reply:
_istioctl_ uses --xds-address or ISTIOCTL_XDS_ADDRESS.